File: app/src/main/java/app/music/base/BaseFragment.java
package app.music.base;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.ViewDataBinding;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;

import app.music.di.component.DaggerFragmentComponent;
import app.music.di.component.FragmentComponent;
import app.music.di.module.FragmentModule;
import app.music.network.DisposableManager;
import app.music.utils.log.InformationLogUtils;
import io.reactivex.disposables.CompositeDisposable;

/**
 * Created by jacky on 3/5/18.
 */

public abstract class BaseFragment<T extends ViewDataBinding> extends Fragment {

    protected T binding;
    protected CompositeDisposable compositeDisposable;
    protected boolean isDestroy;
    private String mLogTag = "BaseFragment";
    protected Boolean mIsVisibleToUser = false;

    @Override
    public void onAttach(Context context) {
        mLogTag = getLogTag();
        InformationLogUtils.INSTANCE.logOnAttach(mLogTag);
        super.onAttach(context);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        InformationLogUtils.INSTANCE.logOnCreate(mLogTag);
        super.onCreate(savedInstanceState);
        initInject();
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        InformationLogUtils.INSTANCE.logOnCreateView(mLogTag);
        compositeDisposable = new CompositeDisposable();
        binding = DataBindingUtil.inflate(inflater, getLayoutId(), container, false);
        initView();
        return binding.getRoot();
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        InformationLogUtils.INSTANCE.logOnActivityCreated(mLogTag);
        super.onActivityCreated(savedInstanceState);
        initData();
//        getBaseActivity().setTitleApp(getTitleApp());
    }

    @Override
    public void onStart() {
        InformationLogUtils.INSTANCE.logOnStart(mLogTag);
        super.onStart();
    }

    @Override
    public void onResume() {
        InformationLogUtils.INSTANCE.logOnResume(mLogTag);
        super.onResume();
    }

    @Override
    public void onPause() {
        InformationLogUtils.INSTANCE.logOnPause(mLogTag);
        super.onPause();
    }

    @Override
    public void onStop() {
        InformationLogUtils.INSTANCE.logOnStop(mLogTag);
        super.onStop();
    }

    @Override
    public void onDestroyView() {
        InformationLogUtils.INSTANCE.logOnDestroyView(mLogTag);
        isDestroy = true;
        if (binding != null) {
            binding.unbind();
        }
        if (compositeDisposable != null) {
            compositeDisposable.dispose();
            compositeDisposable.clear();
        }
        DisposableManager.disposeAll();
        super.onDestroyView();
    }

    @Override
    public void onDestroy() {
        InformationLogUtils.INSTANCE.logOnDestroy(mLogTag);
        super.onDestroy();
    }

    @Override
    public void onDetach() {
        InformationLogUtils.INSTANCE.logOnDetach(mLogTag);
        super.onDetach();
    }

    protected FragmentComponent getFragmentComponent() {
        return DaggerFragmentComponent.builder()
                .appComponent(BaseApplication.getAppComponent())
                .fragmentModule(getFragmentModule())
                .build();
//        return null;
    }

    protected FragmentModule getFragmentModule() {
        return new FragmentModule(this);
    }

    @Override
    public void setUserVisibleHint(boolean isVisibleToUser) {
        super.setUserVisibleHint(isVisibleToUser);
        mIsVisibleToUser = isVisibleToUser;
        if (isResumed()) { // fragment have created
            if (mIsVisibleToUser) {
                onVisible(); // when user switch to another fragment, we can reload data here
            } else {
                onInVisible();
            }
        }
    }

    protected void onVisible() {
        mIsVisibleToUser = true;
    }

    protected void onInVisible() {
        mIsVisibleToUser = false;
    }

    protected void initInject() {

    }

    public abstract int getLayoutId();

    public abstract String getLogTag();

    public abstract void initView();

    public abstract void initData();

    public BaseActivity getBaseActivity() {
        if (getActivity() instanceof BaseActivity) {
            return (BaseActivity) getActivity();
        }
        return null;
    }

    public void replaceFragment(BaseFragment fragment, Bundle mBundle) {
        FragmentTransaction fragmentTransaction
                = getBaseActivity().getSupportFragmentManager().beginTransaction();
        fragmentTransaction.replace(getContainerId(), fragment);
        fragmentTransaction.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN);
        fragmentTransaction.addToBackStack(null);
        fragmentTransaction.commit();
    }

    public int getContainerId() {
        return 0;
    }

    public String getTitleApp() {
        return "";
    }
}

File: app/src/main/java/app/music/base/BaseFragment.kt
package app.music.base

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentTransaction
import app.music.di.component.FragmentComponent
import app.music.di.module.FragmentModule
import app.music.network.DisposableManager
import app.music.utils.log.InformationLogUtils
import io.reactivex.disposables.CompositeDisposable

/**
 * Created by jacky on 3/5/18.
 */

abstract class BaseFragment<T : ViewDataBinding> : Fragment() {

    protected lateinit var binding: T
    var compositeDisposable: CompositeDisposable? = null
    protected var isDestroy: Boolean = false
    private var mLogTag = "BaseFragment"
    protected var mIsVisibleToUser: Boolean = false
    abstract val layoutId: Int
    abstract val logTag: String

    val baseActivity: BaseActivity<*>?
        get() = if (activity is BaseActivity<*>) {
            activity as BaseActivity<*>?
        } else null

    val containerId: Int
        get() = 0

    val titleApp: String
        get() = ""

    override fun onAttach(context: Context) {
        mLogTag = logTag
        InformationLogUtils.logOnAttach(mLogTag)
        super.onAttach(context)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        InformationLogUtils.logOnCreate(mLogTag)
        super.onCreate(savedInstanceState)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        initInject()
        super.onViewCreated(view, savedInstanceState)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        InformationLogUtils.logOnCreateView(mLogTag)
        compositeDisposable = CompositeDisposable()
        binding = DataBindingUtil.inflate(inflater, layoutId, container, false)
        initView()
        return binding.root
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        InformationLogUtils.logOnActivityCreated(mLogTag)
        super.onActivityCreated(savedInstanceState)
        initData()
        //        getBaseActivity().setTitleApp(getTitleApp());
    }

    override fun onStart() {
        InformationLogUtils.logOnStart(mLogTag)
        super.onStart()
    }

    override fun onResume() {
        InformationLogUtils.logOnResume(mLogTag)
        super.onResume()
    }

    override fun onPause() {
        InformationLogUtils.logOnPause(mLogTag)
        super.onPause()
    }

    override fun onStop() {
        InformationLogUtils.logOnStop(mLogTag)
        super.onStop()
    }

    override fun onDestroyView() {
        InformationLogUtils.logOnDestroyView(mLogTag)
        isDestroy = true
        binding.unbind()
        compositeDisposable?.run {
            dispose()
            clear()
        }
        DisposableManager.disposeAll()
        super.onDestroyView()
    }

    override fun onDestroy() {
        InformationLogUtils.logOnDestroy(mLogTag)
        super.onDestroy()
    }

    override fun onDetach() {
        InformationLogUtils.logOnDetach(mLogTag)
        super.onDetach()
    }

    protected fun getFragmentComponent(): FragmentComponent? {
//        return DaggerFragmentComponent.builder()
//                .appComponent(BaseApplication.getAppComponent())
//                .fragmentModule(getFragmentModule())
//                .build()
        return null;
    }

    protected fun getFragmentModule(): FragmentModule {
        return FragmentModule(this)
    }

    override fun setUserVisibleHint(isVisibleToUser: Boolean) {
        super.setUserVisibleHint(isVisibleToUser)
        mIsVisibleToUser = isVisibleToUser
        if (isResumed) { // fragment have created
            if (mIsVisibleToUser) {
                onVisible() // when user switch to another fragment, we can reload data here
            } else {
                onInVisible()
            }
        }
    }

    open fun onVisible() {
        mIsVisibleToUser = true
    }

    open fun onInVisible() {
        mIsVisibleToUser = false
    }

    abstract fun initView()

    abstract fun initData()

    open fun initInject() {

    }

    fun replaceFragment(fragment: BaseFragment<*>, mBundle: Bundle) {
        val fragmentTransaction = baseActivity!!.supportFragmentManager.beginTransaction().apply {
            replace(containerId, fragment)
            setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN)
            addToBackStack(null)
        }
        fragmentTransaction.commit()
    }
}